cluster/kes.go

```diff
@@ -0,0 +1,499 @@
+			Template: corev1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{
+						"app":  deploymentName,
+						"type": "kes",
+					},
+				},
+				Spec: kesPodSpec,
+			},
+			Strategy: appsv1.DeploymentStrategy{
+				Type: appsv1.RecreateDeploymentStrategyType,
+			},
+		},
+	}
+}
+
+func createNewKesDeployment(clientset *kubernetes.Clientset, deploymentName string, kesSecretsNames map[string]string) <-chan struct{} {
+	doneCh := make(chan struct{})
+	go func() {
+		defer close(doneCh)
+		factory := informers.NewSharedInformerFactory(clientset, 0)
+		deploymentInformer := factory.Apps().V1().Deployments().Informer()
+		deploymentInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+			UpdateFunc: func(oldObj, newObj interface{}) {
+				deployment := newObj.(*appsv1.Deployment)
+				if deployment.Name == deploymentName && len(deployment.Status.Conditions) > 0 && deployment.Status.Conditions[0].Status == "True" {
+					log.Println("kes deployment created correctly")
+					return
+				}
+			},
+		})
+		go deploymentInformer.Run(doneCh)
+
+		//Creating nginx-resolver deployment with new rules
+		kesDeployment := getNewKesDeployment(deploymentName, kesSecretsNames)
+		_, err := appsV1API(clientset).Deployments("default").Create(&kesDeployment)
+		if err != nil {
+			log.Println(err)
+			return
+		}
+	}()
+	return doneCh
+}
+
+func createNewKesService(clientset *kubernetes.Clientset, serviceName string) <-chan struct{} {
+	doneCh := make(chan struct{})
+	go func() {
+		defer close(doneCh)
+		factory := informers.NewSharedInformerFactory(clientset, 0)
+		serviceInformer := factory.Core().V1().Services().Informer()
+		serviceInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				service := obj.(*v1.Service)
+				if service.Name == serviceName {
+					log.Println("kes service created correctly")
+					return
+				}
+			},
+		})
+
+		go serviceInformer.Run(doneCh)
+
+		pgSvc := v1.Service{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: serviceName,
+				Labels: map[string]string{
+					"name": serviceName,
+				},
+			},
+			Spec: v1.ServiceSpec{
+				Ports: []v1.ServicePort{
+					{
+						Port: 7373,
+					},
+				},
+				Selector: map[string]string{
+					"app": serviceName,
+				},
+			},
+		}
+		_, err := clientset.CoreV1().Services("default").Create(&pgSvc)
+		if err != nil {
+			log.Println(err)
+			return
+		}
+	}()
+	return doneCh
+}
+
+type KeyPair struct {
+	cert         string
+	certIdentity string
+	key          string
+}
+
+func generateKeyPair(name string) chan *KeyPair {
+	doneCh := make(chan *KeyPair)
+	go func() {
+		defer close(doneCh)
+		public, private, err := ed25519.GenerateKey(rand.Reader)
+		if err != nil {
+			log.Println(err)
+			return
+		}
+		serialNumberLimit := new(big.Int).Lsh(big.NewInt(1), 128)
+		serialNumber, err := rand.Int(rand.Reader, serialNumberLimit)
+		if err != nil {
+			log.Println(err)
+			return
+		}
+		now := time.Now()
```

Comment:
we should proably use time.Now().UTC(), we created `UTCNow()` on utilis.go if needed.

Reply:
x509.Certificate doesn't care about the timezone as long as the provided value is a Time object, however I'm also a fan of UTC, I'll ask @aead if this could have some implications when generating the keypair.